accounts/abi/bind/dep_tree.go

```diff
@@ -0,0 +1,240 @@
+package bind
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"regexp"
+	"strings"
+)
+
+// DeploymentParams represents parameters needed to deploy a
+// set of contracts.  It takes an optional override
+// list to specify contracts/libraries that have already been deployed on-chain.
+type DeploymentParams struct {
+	contracts []*MetaData
+	// map of solidity library pattern to constructor input.
+	inputs map[string][]byte
+	// Overrides is an optional map of pattern to deployment address.
+	// Contracts/libraries that refer to dependencies in the override
+	// set are linked to the provided address (an already-deployed contract).
+	overrides map[string]common.Address
+}
+
+// NewDeploymentParams instantiates a DeploymentParams instance.
+func NewDeploymentParams(contracts []*MetaData, inputs map[string][]byte, overrides map[string]common.Address) *DeploymentParams {
+	return &DeploymentParams{
+		contracts,
+		inputs,
+		overrides,
+	}
+}
+
+// DeploymentResult encapsulates information about the result of the deployment
+// of a set of contracts: the pending deployment transactions, and the addresses
+// where the contracts will be deployed at.
+type DeploymentResult struct {
+	// map of contract library pattern -> deploy transaction
+	Txs map[string]*types.Transaction
+	// map of contract library pattern -> deployed address
+	Addrs map[string]common.Address
+}
+
+// Accumulate merges two DeploymentResult objects together.
+func (d *DeploymentResult) Accumulate(other *DeploymentResult) {
+	for pattern, tx := range other.Txs {
+		d.Txs[pattern] = tx
+	}
+	for pattern, addr := range other.Addrs {
+		d.Addrs[pattern] = addr
+	}
+}
+
+// depTreeBuilder turns a set of unlinked contracts libraries into a set of one
+// or more dependency trees.
+type depTreeBuilder struct {
+	overrides map[string]common.Address
```

Comment:
I think the whole handling of overrides is a bit flaky. You hae three things: 
- A map of `overrides` here in `depTreeBuilder`
- Those get individually assigned as `overrideAddr` in individual `depTreeNode` instances
- And eventually used as `linkAddr`, used instead of the `d.deployedAddrs` in `depTreeDeployer`. 

I think a better model would be to consider the overrides as deployments that already happened. Would it be possible to just transfer the `overrides` from `depTreeBuilder` into `d.deployedAddrs` in `depTreeDeployer`? And it would simply not travese children that are already deployed (already has an addr in the map). 

Then we would not need the field `overrideAddr` i the `depTreeNode`.

Reply:
Originally, I didn't have the concept of a node having an `overrideAddr` and simply didn't recurse on any nodes that were in the override set.

However, there's a problem with that approach.  Let's say we have a dependency graph:
```
A    B.   C.   D
 \.   /      /.    /.
   E.     /.     /
     \.  /.     /
           F (contract)
```

If I specify `E` as an override, and pass F's dependencies (the flat set of everything direct/indirect is emitted in the bindings):  The deployer code skips recursiing on `E`, and then incorrectly assumes that `A`/`B` (being disconnected from the rest of the dependency graph) should be deployed.

Reply:
Well, why/when do you operate on the flat set? The dependices of `F` are `[E,C,D]`. 
And in this case, `E` is already deployed. Sounds like an issue elsewhere if A,B pops up and cause problems

Reply:
I think the semantics are simpler than what you have. It's not about overrides, it's about already deployed. 

It should not matter if your code just deployed it, or it was deployed a year ago and the user just tells you about it. IMO it's the same thing, but in the implementation you differentiate them, and call one of them "overrides".

Reply:
> Well, why/when do you operate on the flat set?

I thought it would be simpler at some point.  Clearly it's not, and we should just generate bindings where each contract's `MetaData` object contains pointers to its direct dependencies.

Reply:
> IMO it's the same thing

not quite:  we need to distinguish between contracts that were previously-deployed (overrides) and contracts that have been deployed by the current/last call to `LinkAndDeploy`.  The reason being that we want to return the set of contracts that were just deployed to the API user, and not include the overrides.

Reply:
> The reason being that we want to return the set of contracts that were just deployed to the API user, and not include the overrides.

Sure, but that can probably be done pretty easily. For the deployed ones, you have a tx. For the "already deployed earlier", you don't. So that is a pretty good distinguishing marker